Add wrapper functionality & bridge integration

loic1 opened the pull request:
Changes:
- Update FlowEVM TopShot contract code:
  - Extend `ERC721Wrapper` OpenZeppelin library
  - Remove `ERC721URIStorage` extension (per-token metadata storage isn't currently needed) and use `baseTokenURI` contract state variable instead to build metadata endpoint url from `tokenURI` with token ID
  - Extend `BridgePermissionsUpgradeable` to prevent bridge onboarding before bridge upgrade
  - Extend `CrossVMBridgeFulfillmentUpgradeable` to allow upgraded bridge to fulfill to EVM (i.e., mint or transfer)
- Add Cadence bridging & wrapping txs

loic1 commented on evm-bridging/src/BridgedTopShotMoments.sol:
```diff
@@ -19,43 +21,85 @@ import {ITransferValidator721} from "./interfaces/ITransferValidator.sol";
     }
 
+    /**
+     * @dev Initializes the contract.
```
Added contract init params: `underlyingNftContractAddress`, `vmBridgeAddress`, `baseTokenURI_`

loic1 commented on evm-bridging/src/BridgedTopShotMoments.sol:
```diff
@@ -73,31 +117,33 @@ contract BridgedTopShotMoments is
         return _customSymbol;
     }
 
-    function safeMint(address to, uint256 tokenId, string memory uri) public onlyOwner {
```
Removed `safeMint` - minting will only be possible as part of:
1. Wrapper functionality
2. Bridge fulfillment (when contract is onboarded to upgraded bridge)

loic1 commented on evm-bridging/src/interfaces/IBridgePermissions.sol:
```diff
@@ -0,0 +1,16 @@
+
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+
+interface IBridgePermissions is IERC165 {
```
From [flow-evm-bridge](https://github.com/onflow/flow-evm-bridge/blob/main/solidity/src/interfaces/IBridgePermissions.sol) repo

loic1 commented on evm-bridging/src/interfaces/ICrossVMBridgeFulfillment.sol:
```diff
@@ -0,0 +1,6 @@
+pragma solidity 0.8.24;
+
+interface ICrossVMBridgeFulfillment {
```
Interface conformance expected to be needed to onboard a custom contract in upcoming upgraded bridge

loic1 commented on evm-bridging/src/lib/BridgePermissionsUpgradeable.sol:
```diff
@@ -0,0 +1,53 @@
+ * Note: The Flow VM bridge checks for permissions at asset onboarding. If your asset has already
+ * been onboarded, setting `permissions` to `false` will not affect movement between VMs.
+ */
+abstract contract BridgePermissionsUpgradeable is Initializable, ERC165Upgradeable, IBridgePermissions {
```
From [flow-evm-bridge](https://github.com/onflow/flow-evm-bridge/blob/main/solidity/src/interfaces/IBridgePermissions.sol) repo

loic1 commented on evm-bridging/src/lib/CrossVMBridgeCallableUpgradeable.sol:
```diff
@@ -0,0 +1,52 @@
+ * @dev A base contract intended for use in implementations on Flow, allowing a contract to define
+ *      access to the Cadence X EVM bridge on certain methods.
+ */
+abstract contract CrossVMBridgeCallableUpgradeable is ContextUpgradeable {
```
Adapted for upgradable contracts from  [flow-evm-repo draft PR](https://github.com/onflow/flow-evm-bridge/pull/168/files#diff-d5335bccc50b538dd4a4bf2bbadda5300ad0865271ef50aa421411d606a1ca44R12) - the functionality is what's currently expected to register a custom contract in upcoming upgraded bridge

loic1 commented on evm-bridging/src/lib/CrossVMBridgeFulfillmentUpgradeable.sol:
```diff
@@ -0,0 +1,66 @@
+import {ERC165Upgradeable} from "@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable.sol";
+import {ICrossVMBridgeFulfillment} from "../interfaces/ICrossVMBridgeFulfillment.sol";
+
+abstract contract CrossVMBridgeFulfillmentUpgradeable is CrossVMBridgeCallableUpgradeable, ERC165Upgradeable, ERC721Upgradeable {
```
Adapted for upgradable contracts from [flow-evm-repo draft PR](https://github.com/onflow/flow-evm-bridge/pull/168/files#diff-77f2f2408b4f1cebb1fb2f47739584dbaefe42648e5e86b08d15febc34d7ab3dR7)  - the functionality is what's currently expected to register a custom contract in upcoming upgraded bridge

loic1 commented on contracts/TopShot.cdc:
```diff
@@ -1717,22 +1725,34 @@ access(all) contract TopShot: NonFungibleToken {
     }
 
+    // resolveCrossVMPointerView resolves the CrossVMPointer view
+    // access(all) view fun resolveCrossVMPointerView(): MetadataViews.CrossVMPointer {
```
Commented out until [this PR](https://github.com/onflow/flow-nft/pull/243/files#diff-c7956c28b08398abbebc13e67db6c3502f28e39764c642bef07f95b19ee95d4cR22)'s changes are merged and pushed on-chain - this new metadata view should be returned both from contract and nft metadata views; this should provide the custom contract FlowEVM address

loic1 commented on evm-bridging/cadence/transactions/bridge_nfts_to_evm_and_wrap.cdc:
```diff
@@ -0,0 +1,189 @@
+import "FlowEVMBridgeConfig"
+import "FlowEVMBridgeUtils"
+
+/// Bridges NFTs with provided IDs from Cadence to EVM and wraps them in a wrapper ERC721
```
All-in-one Cadence tx to bridge NFTs from Cadence to EVM and make EVM contract calls to approve wrapper erc721 contract as operator and wrap the bridged NFTs to wrapper erc721 NFTs

loic1 commented on evm-bridging/cadence/transactions/bridge_nfts_to_evm_and_wrap.cdc:
```diff
@@ -0,0 +1,189 @@
+/// @param nftIDs: Array of IDs of the NFTs to wrap
+///
+transaction(
+    wrapperERC721Address: String,
```
This parameter should not be needed anymore once we can get the contract address from new metadata view on-chain

loic1 commented:
Updated bridge interface & extension to match latest from https://github.com/onflow/flow-evm-bridge/pull/168 (commit `ef81bae`)

vdojnov commented on evm-bridging/cadence/transactions/bridge_nfts_to_evm_and_wrap.cdc:
```diff
@@ -0,0 +1,189 @@
+    let res = coa.call(
+        to: wrapperAddress,
+        data: EVM.encodeABIWithSignature("getCadenceIdentifier()", []),
+        gasLimit: 100_000,
```
Question for understanding, why do we sometimes put 100k vs 400k like below?

loic1 replied:
Used 100k for all the calls initially but then ran into `intrinsic gas too low` error - this call is read-only vs the other ones actually update contract values and have multiple checks before doing so that also cost some gas

vdojnov commented on evm-bridging/script/Deploy.s.sol:
```diff
@@ -18,19 +18,24 @@ contract DeployScript is Script {
         address owner = msg.sender;
         string memory name = "NBA Top Shot";
         string memory symbol = "TOPSHOT";
+        string memory baseTokenURI = "https://api.cryptokitties.co/tokenuri/";
```
The current prod one is `https://metadata-api.production.studio-platform.dapperlabs.com/v1/topshot/moment/{{momentID}}` Maybe we can put this as a place holder for now? But if we want a nice public facing URL we can work with SRE

loic1 commented on evm-bridging/README.md:
````diff
@@ -1,71 +1,133 @@
 
 ```sh
 cp .env.example.testnet .env
+# Add your account details to .env and source it
````
this one `.env.flowevm.testnet.example` is just a template for testnet where we need to populate account address & private key to send txs with `cast send`

judezhu replied:
i mean `.env.example.testnet` seems to be the wrong name. `.env.flowevm.testnet.example` is the correct name

loic1 replied:
oh I see, fixed - thank you!

loic1 commented on evm-bridging/README.md:
```diff
@@ -1,71 +1,133 @@
 
-## Execute Queries
+# Approve operator for all NFTs
+cast send $DEPLOYED_PROXY_CONTRACT_ADDRESS --rpc-url $RPC_URL --private-key <private-key> --legacy "setApprovalForAll(address,bool)" <operator-address> <true>
```
`setApprovalForAll` gives permission for the operator to transfer any NFTs owned by the caller, unlike `approve`, which gives permission for specific nft id

loic1 commented on evm-bridging/cadence/transactions/bridge_nfts_to_evm_and_wrap.cdc:
```diff
@@ -0,0 +1,189 @@
+        value: EVM.Balance(attoflow: 0)
+    )
+
+    assert(res.status == EVM.Status.successful, message: "Call to get underlying ERC721 address failed")
```
This one doesn't fail silently, the one that should is from [the flow tx PR](https://github.com/Outblock/FRW-web-next/pull/156/files#diff-878c41f748013c56b4979011cbeaaf059ed5ffba6a6b4d5b4b0255525255eb2eR157):
```
        // Get the project-defined ERC721 address if it exists
        if let crossVMPointer = viewResolver.resolveContractView(
                resourceType: self.nftType,
                viewType: Type<MetadataViews.CrossVMPointer>()
        ) as! MetadataViews.CrossVMPointer? {
            // Wrap NFTs if necessary
            if let underlyingAddress = getUnderlyingERC721Address(
                self.coa,
                crossVMPointer.evmContractAddress
            ) {
                // Approve contract to withdraw underlying NFTs from signer's coa
                call(self.coa, underlyingAddress,
                    functionSig: "setApprovalForAll(address,bool)",
                    args: [crossVMPointer.evmContractAddress, true]
                )

                // Wrap NFTs with provided IDs
                call(self.coa, crossVMPointer.evmContractAddress,
                    functionSig: "depositFor(address,uint256[])",
                    args: [self.coa.address(), ids]
                )
            }
        }
```

loic1 commented on evm-bridging/cadence/transactions/admin/set_up_royalty_management.cdc:
```diff
@@ -0,0 +1,57 @@
+        // Set royalty info
+        call(coa, EVM.addressFromString(erc721C),
+            functionSig: "setRoyaltyInfo((address,uint96))",
+            args: [[EVM.addressFromString(validator), royaltyBasisPoints]]
```
Yes! fixed, thank you!

loic1 commented on evm-bridging/cadence/transactions/bridge_nfts_to_evm_and_wrap.cdc:
```diff
@@ -0,0 +1,189 @@
+        call(self.coa, underlyingAddress,
+            functionSig: "setApprovalForAll(address,bool)",
+            args: [wrapperAddress, true]
+        )
```
@judezhu Good question! Definitely seems good practice to only approve the specific NFTs that need to be transferred but I think approve for all is reasonable in this case because the operator we are approving is erc721c/wrapper, which is a trusted contract and it will be the only official evm contract that moments will be able to be bridged to once the bridge is updated and we onboard the contract to it - does that make sense? 👀

loic1 replied:
we can save gas by avoiding repeated `approve` calls for each of the NFTs passed to `depositFor` for wrapping

loic1 replied:
@sisyphusSmiling Added call to `setApprovalForAll`, thank you!

Re: uint encoded
Was able to test that the max UInt64 value gets encoded and decoded with no problem, see [test tx](https://github.com/dapperlabs/nba-smart-contracts/pull/282/files#diff-284ede1095651956c342ca4ec9ef93eb9f7df9351638011dbeb023ef3aa75bea) calling [test contract](https://github.com/dapperlabs/nba-smart-contracts/pull/282/files#diff-9179a5a541ce4c88af86f9d251f07ff6b182902eae98418be4550d2e7b49afa2R5) - is this what you meant we should confirm?

sisyphusSmiling replied:
That makes sense, thanks for testing. I'd also asked in Discord and the Dete mentioned that all encoded values use 256 bit width, so it makes sense that it would work.

loic1 replied:
Gotcha, thank you for sharing!

loic1 commented on evm-bridging/test/BridgedTopShotMoments.t.sol:
```diff
@@ -5,25 +5,74 @@ import {Test} from "forge-std/Test.sol";
+
+        // Set NFT contract initialization parameters
+        underlyingNftContractAddress = address(underlyingNftContract);
+        vmBridgeAddress = address(0x67890);
```
@sisyphusSmiling could you confirm `vmBridgeAddress` should be the address specified in https://github.com/onflow/flow-evm-bridge/blob/main/README.md?plain=1#L24-L25? 🙏

sisyphusSmiling replied:
Yes, that's correct.

zhangchiqing commented on evm-bridging/src/lib/CrossVMBridgeERC721FulfillmentUpgradeable.sol:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /**
+     * @dev Returns whether the token is exists or not defined positively by whether the owner of
```
We probably need to mention the scope (where) of the `exists`. Does it exist in the VM bridge or actually exist in the original contract. 

Please correct me if wrong. My guess is when it return `false`, it means it just doesn't exist in the bridged VM yet, and it can't tell whether it exists in the NBA contract in Cadence, right?

For instance, if Account A owns a topshot NFT #100 (token ID: 100) in cadence, now the token has not moved to Flow EVM yet, so the `exists(100)` would return `false`. Then, after user A moves this token to Flow EVM, then `exists(100)` would return `true`. Is that correct?

I'm also a bit confused by the "VM" in `CrossVM`. In my mind, there are many VMs here: Flow VM (the flow native VM), Flow EVM (the EVM in Flow), Ethereum EVM (Ethereum's native EVM), Optimum EVM (Layer 2 EVM). 

Does CrossVMBridge mean the bridge between Flow VM and Flow EVM? Or means the bridge between Flow EVM, Ethereum EVM Optimum EVM? If is later, maybe we could call it CrossEVMBridge to be more specific. 

If CrossVMBridge actually mean CrossEVMBridge, then the `exists` would be a bit tricky to implement I guess, because it requires an async cross chain call to know if the token exist in the Flow native VM, right?

sisyphusSmiling replied:
> For instance, if Account A owns a topshot NFT #100 (token ID: 100) in cadence, now the token has not moved to Flow EVM yet, so the exists(100) would return false. Then, after user A moves this token to Flow EVM, then exists(100) would return true. Is that correct?

Yes, that's correct. 

> Does CrossVMBridge mean the bridge between Flow VM and Flow EVM? Or means the bridge between Flow EVM, Ethereum EVM Optimum EVM? If is later, maybe we could call it CrossEVMBridge to be more specific.

When we say "VM" here it's in reference to environments on Flow, so "cross-vm" is meant to imply movement and/or association across Cadence & EVM as in "across VMs". This is maybe a bit of a semantic failing on my part motivated by a desire to adapt the idea of "cross-chain" bridge to our use case, which is more of a "cross-vm" bridge implementation.

I've seen `exists` in other ERCs (e.g.`ERC1155Supply`) and it was formerly internal to OZ's ERC721 implementation, so here is implemented as port of those to see if an ERC721 token with the id exists. Given the directionality of calls between Cadence -> EVM, we wouldn't be able to inspect Cadence to determine existence on top of the difficulty of finding the resource given the distributed storage model of Cadence NFTs.

zhangchiqing replied:
Thanks for clarifying. Then I suggest to mention it in the comments.
```suggestion
     * @dev Returns whether the token exists in the bridge or not defined positively by whether the owner of
```

And just for my understanding:

I wonder if `exists` and `isEscrowed` return different value? It seems to me, their implementation are different. To me, if a token exists in the bridge, then its own must be the `vmBridgeAddress`, right? would the `_ownerOf(id)` return a different than the bridge address? 

I think `_ownerOf(id)` can either be 0 or the bridge address, if it's trying to store the real owner, eventually that value will be stale, right?

loic1 replied:
@zhangchiqing `exists` should return true for any existing NFT that hasn't been burned (sent to zero address), while `isEscrowed` should return true only if the NFT's current owner is the bridge address

> It seems to me, if a token exists in the bridge, then its own must be the vmBridgeAddress, right? would the _ownerOf(id) return a different value than the bridge address?

If a NFT exists in the bridge, `isEscrowed` should return true and so should `exists` because the NFT isn't burned

> I think _ownerOf(id) can either be 0 or the bridge address, if it's trying to store the real owner, eventually that value will be stale, right?

Sorry I'm not completely sure what you mean here. `_ownerOf` can be any evm address, the libary is only extending the erc721 contract, which is where `_ownerOf` is defined

sisyphusSmiling commented on evm-bridging/test/BridgedTopShotMoments.t.sol:
Great to see tests here. I'd like to see the Cadence transactions tested as well since those will be critical.

loic1 replied:
Added Cadence transaction testing in https://github.com/dapperlabs/nba-smart-contracts/pull/282 for royalty admin operations.

We are testing updated bridging txs for wrapping/unwrapping on testnet for now. Local testing for these would be nice to have but would also require setting up a test environment with flow evm bridge dependencies. This might also make sense to add generic tests in flow-evm-bridge repo if the bridge adds built-in support for wrapping/unwrapping erc721s as part of the changes for custom association FLIP.

sisyphusSmiling replied:
Yeah, definitely will get those added once the functionality is built out

loic1 commented on evm-bridging/cadence/transactions/admin/set_up_royalty_management_args.json:
Yes this is just placeholder, readme indicates actual arguments should be populated

loic1 commented on evm-bridging/README.md:
```diff
@@ -1,76 +1,139 @@
+
+2. **Bridge Integration**
+   - Wrapper functionality for ERC721s from bridged-deployed contract
+   - Cross-VM compatibility for Flow ↔ EVM bridging (after bridge upgrade allowing custom associations, and after contract is onboarded to the bridge)
```
Made the suggested change, thank you!

loic1 commented on evm-bridging/cadence/transactions/admin/set_up_royalty_management.cdc:
```diff
@@ -0,0 +1,57 @@
+/// @param erc721C - The address of the ERC721 contract
+/// @param validator - The address of the validator contract
+/// @param royaltyRecipient - The address of the royalty recipient
+/// @param royaltyBasisPoints - The royalty basis points (0-10000)
```
Made the suggested change, thank you!

loic1 commented on evm-bridging/cadence/transactions/transfer_erc721s_to_evm_address.cdc:
```diff
@@ -0,0 +1,56 @@
+import "FlowToken"
+import "EVM"
+
+/// Transfers an NFT from the signer's COA to an EVM address
```
Made the suggested change, thank you!

loic1 commented on evm-bridging/cadence/transactions/bridge_nfts_to_evm_and_wrap.cdc:
```diff
@@ -0,0 +1,197 @@
+    _ contractAddr: EVM.EVMAddress,
+    functionSig: String,
+    args: [AnyStruct],
+) {
```
Added the `EVM.EVMResult` return and validation of `depositFor` and `withdrawTo` boolean returns in all txs where the functions are called - thank you for the suggestion!

loic1 commented on evm-bridging/cadence/transactions/unwrap_nfts.cdc:
```diff
@@ -0,0 +1,80 @@
+        mustCall(self.coa, wrapperAddress,
+            functionSig: "withdrawTo(address,uint256[])",
+            args: [self.coa.address(), nftIDs]
+        )
```
Added check as said in previous response

loic1 commented on evm-bridging/cadence/transactions/unwrap_nfts_and_bridge_from_evm.cdc:
```diff
@@ -0,0 +1,188 @@
+        mustCall(self.coa, EVM.addressFromString(wrapperERC721Address),
+            functionSig: "withdrawTo(address,uint256[])",
+            args: [self.coa.address(), nftIDs]
+        )
```
Added second call to set back approval for all to false (see previous comment)

loic1 commented on evm-bridging/cadence/transactions/wrap_nfts.cdc:
```diff
@@ -0,0 +1,80 @@
+        let underlyingAddress = getUnderlyingERC721Address(self.coa, wrapperAddress)
+
+        // Approve contract to withdraw underlying NFTs from signer's coa
+        mustCall(self.coa, underlyingAddress,
+            functionSig: "setApprovalForAll(address,bool)",
+            args: [wrapperAddress, true]
+        )
```
Added second call to set back approval for all to false (see earlier comment)

loic1 commented on contracts/TopShot.cdc:
```diff
@@ -1717,22 +1725,34 @@ access(all) contract TopShot: NonFungibleToken {
+    // resolveCrossVMPointerView resolves the CrossVMPointer view
+    // access(all) view fun resolveCrossVMPointerView(): MetadataViews.CrossVMPointer {
+    //     return MetadataViews.CrossVMPointer(
+    //         cadenceType: Type<@TopShot>(),
```
Added the passthru, thank you for the suggestion!

loic1 commented on evm-bridging/cadence/scripts/get_underlying_erc721_address.cdc:
```diff
@@ -0,0 +1,33 @@
+
+/// Returns the hex encoded address of the underlying ERC721 contract
+///
+access(all) fun main(flowNftAddress: Address, wrapperERC721Address: String): String? {
```
Cleaned up unnecessary script in https://github.com/dapperlabs/nba-smart-contracts/pull/282

loic1 commented on evm-bridging/cadence/scripts/get_underlying_erc721_address.cdc:
```diff
@@ -0,0 +1,33 @@
+access(all) fun getUnderlyingERC721Address(
+    _ coa: auth(EVM.Call) &EVM.CadenceOwnedAccount,
+    _ wrapperAddress: EVM.EVMAddress
+): EVM.EVMAddress {
```
Cleaned up unnecessary script in https://github.com/dapperlabs/nba-smart-contracts/pull/282

loic1 commented on evm-bridging/script/Deploy.s.sol:
```diff
@@ -18,19 +18,24 @@ contract DeployScript is Script {
         string memory symbol = "TOPSHOT";
+        string memory baseTokenURI = "https://api.cryptokitties.co/tokenuri/";
         string memory cadenceNFTAddress = "877931736ee77cff";
         string memory cadenceNFTIdentifier = "A.877931736ee77cff.NFT";
```
Fyi, working on the above in https://github.com/dapperlabs/nba-smart-contracts/pull/282

loic1 commented on evm-bridging/src/lib/BridgePermissionsUpgradeable.sol:
```diff
@@ -0,0 +1,53 @@
+     */
+    function _setPermissions(bool permissions) internal {
+        _permissions = permissions;
+        emit PermissionsUpdated(permissions);
```
@zhangchiqing made the change, thanks for the suggestion!

Fyi this is adapted from a flow-evm-bridge library (to make it compatible with upgradable contracts) not a new library (https://github.com/dapperlabs/nba-smart-contracts/pull/281#discussion_r1949940204)

cc @sisyphusSmiling for vis and potentially update flow-evm-bridge library

loic1 commented on evm-bridging/src/BridgedTopShotMoments.sol:
```diff
@@ -19,44 +21,88 @@ import {ITransferValidator721} from "./interfaces/ITransferValidator.sol";
+    // Event declarations
     event RoyaltyInfoUpdated(address receiver, uint256 bps);
+    event ContractURIUpdated();
+    event MetadataUpdate(uint256 tokenId);
```
Added the IERC4906 implementation, thanks!

loic1 commented on evm-bridging/cadence/transactions/admin/set_up_royalty_management.cdc:
```diff
@@ -0,0 +1,57 @@
+
+        // Set royalty info
+        mustCall(coa, EVM.addressFromString(erc721C),
+            functionSig: "setRoyaltyInfo((address,uint96))",
```
The function signature does show as `setRoyaltyInfo((address,uint96))` in decoded input data on blockchain explorer

Re: uint96 encoding/decoding
Looks like we clarified with testing and discord conversation that it should not be an issue
https://github.com/dapperlabs/nba-smart-contracts/pull/281#discussion_r1960562321

loic1 commented on evm-bridging/cadence/transactions/admin/set_up_royalty_management.cdc:
```diff
@@ -0,0 +1,57 @@
+        // Set royalty info
+        mustCall(coa, EVM.addressFromString(erc721C),
+            functionSig: "setRoyaltyInfo((address,uint96))",
+            args: [[EVM.addressFromString(royaltyRecipient), royaltyBasisPoints]]
```
Fixed, thanks

loic1 commented:
Key Updates: 
- Added `EVMPointer` view support for newly deployed `CrossVMMetadataViews` (uncommented previously pushed changes)
  - Uses placeholder for custom erc721 address until the actual testnet and mainnet contracts are deployed
  - Added local test dependency handling
- Updated txs to wrap/unwrap only when applicable and to transfer to optional provided evm address (with separate util methods)
- Added tx fixes & improvements, `withdrawTo`/`depositFor` return checks, `safeTransferFrom` ownership checks  - based on PR feedback + testing

vdojnov commented on contracts/TopShot.cdc:
```diff
@@ -1717,18 +1727,26 @@ access(all) contract TopShot: NonFungibleToken {
+                    cadenceType: Type<@TopShot.NFT>(),
+                    cadenceContractAddress: self.account.address,
+                    // TODO: Replace with actual EVM contract address
+                    evmContractAddress: EVM.addressFromString("0x1234565789012345657890123456578901234565"),
```
Should we add a 'variable' thing here and remove the comment?

```suggestion
                    evmContractAddress: EVM.addressFromString("0xEVMCONTRACTADDRESS"),
```

judezhu commented on evm-bridging/cadence/transactions/wrap_nfts.cdc:
```diff
@@ -0,0 +1,89 @@
+    return decodedResult[0] as! EVM.EVMAddress
+}
+
+/// Calls a function on an EVM contract from provided coa
```
so we decided not to deploy an util contract for the `mustCall` function?

loic1 replied:
Yea it would be best achieved by holding the contract in a trusted neutral entity like the bridge account or a keyless account given the privileged access to COA, but it doesn't seem really worth it at this point as the wrapping/unwrapping logic is needed only temporarily until the bridge gains support for custom associations and unwrapping logic within its own context

judezhu reviewed (approved):
LGTM! left one question
